File: research/biomni/knowhow/workflows/scrnaseq-scanpy-core-analysis/scripts/plot_dimreduction.py
"""
============================================================================
DIMENSIONALITY REDUCTION VISUALIZATION
============================================================================

This script creates publication-quality plots of dimensionality reductions.

Functions:
  - plot_umap_clusters(): UMAP colored by clusters
  - plot_clustering_comparison(): Compare multiple resolutions
  - plot_feature_umap(): UMAP colored by gene expression or QC metrics
  - plot_umap_with_plotnine(): Create UMAP plots using plotnine

Usage:
  from plot_dimreduction import plot_umap_clusters, plot_feature_umap
  plot_umap_clusters(adata, cluster_key='leiden_0.8', output_dir='results/umap')
  plot_feature_umap(adata, features=['CD3D', 'CD14'], output_dir='results/umap')
"""

from pathlib import Path
from typing import List, Optional, Union

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd


def _save_plot(fig: plt.Figure, base_path: Union[str, Path], dpi: int = 300) -> None:
    """
    Save plot in both PNG and SVG formats with graceful fallback.

    Parameters
    ----------
    fig : matplotlib.figure.Figure
        Figure object to save
    base_path : str or Path
        Base path for output files (without extension)
    dpi : int, optional
        Resolution for PNG (default: 300)

    Returns
    -------
    None
        Saves files to disk
    """
    base_path = Path(base_path)

    # Always save PNG
    png_path = base_path.with_suffix('.png')
    try:
        fig.savefig(png_path, dpi=dpi, bbox_inches='tight', format='png')
        print(f"  Saved: {png_path}")
    except Exception as e:
        print(f"  Warning: PNG export failed: {e}")

    # Always try SVG
    svg_path = base_path.with_suffix('.svg')
    try:
        fig.savefig(svg_path, bbox_inches='tight', format='svg')
        print(f"  Saved: {svg_path}")
    except Exception as e:
        print(f"  (SVG export failed, PNG available)")


def plot_umap_clusters(
    adata: 'AnnData',
    cluster_key: str = 'leiden_0.8',
    output_dir: Union[str, Path] = ".",
    figsize: tuple = (8, 6),
    palette: Optional[str] = None
) -> None:
    """
    Create UMAP plot colored by clusters.

    Parameters
    ----------
    adata : AnnData
        AnnData object with UMAP
    cluster_key : str, optional
        Cluster column in adata.obs (default: 'leiden_0.8')
    output_dir : str or Path, optional
        Output directory for plot (default: ".")
    figsize : tuple, optional
        Figure size (default: (8, 6))
    palette : str, optional
        Color palette (default: None, uses scanpy default)

    Returns
    -------
    None
        Saves plot to output_dir
    """
    import scanpy as sc

    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    if 'X_umap' not in adata.obsm:
        print("Warning: UMAP not found. Run run_umap_reduction first.")
        return

    if cluster_key not in adata.obs.columns:
        print(f"Warning: {cluster_key} not found in adata.obs")
        return

    print(f"Plotting UMAP colored by {cluster_key}...")

    fig, ax = plt.subplots(figsize=figsize)
    sc.pl.umap(
        adata,
        color=cluster_key,
        palette=palette,
        legend_loc='on data',
        legend_fontsize='x-small',
        legend_fontoutline=2,
        frameon=False,
        show=False,
        ax=ax
    )

    output_file = output_dir / f"umap_{cluster_key}"
    fig = plt.gcf()
    _save_plot(fig, output_file, dpi=300)
    plt.close()


def plot_clustering_comparison(
    adata: 'AnnData',
    resolutions: List[float] = [0.4, 0.6, 0.8, 1.0],
    output_dir: Union[str, Path] = ".",
    figsize: tuple = (16, 4)
) -> None:
    """
    Compare clustering at multiple resolutions.

    Parameters
    ----------
    adata : AnnData
        AnnData object with UMAP and multiple clustering results
    resolutions : list of float, optional
        Resolutions to compare (default: [0.4, 0.6, 0.8, 1.0])
    output_dir : str or Path, optional
        Output directory for plot (default: ".")
    figsize : tuple, optional
        Figure size (default: (16, 4))

    Returns
    -------
    None
        Saves plot to output_dir
    """
    import scanpy as sc

    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    if 'X_umap' not in adata.obsm:
        print("Warning: UMAP not found. Run run_umap_reduction first.")
        return

    print(f"Plotting clustering comparison for {len(resolutions)} resolutions...")

    # Check which resolutions exist
    cluster_keys = [f'leiden_{res}' for res in resolutions]
    existing_keys = [key for key in cluster_keys if key in adata.obs.columns]

    if len(existing_keys) == 0:
        print("Warning: No clustering results found for specified resolutions")
        return

    sc.pl.umap(
        adata,
        color=existing_keys,
        ncols=len(existing_keys),
        legend_loc='on data',
        legend_fontsize='xx-small',
        legend_fontoutline=2,
        frameon=False,
        show=False
    )

    output_file = output_dir / "umap_resolution_comparison"
    fig = plt.gcf()
    _save_plot(fig, output_file, dpi=300)
    plt.close()


def plot_feature_umap(
    adata: 'AnnData',
    features: List[str],
    use_raw: bool = False,
    layer: Optional[str] = None,
    output_dir: Union[str, Path] = ".",
    figsize: Optional[tuple] = None,
    ncols: int = 3,
    cmap: str = 'viridis'
) -> None:
    """
    Create UMAP plots colored by gene expression or metadata.

    Parameters
    ----------
    adata : AnnData
        AnnData object with UMAP
    features : list of str
        Features to plot (genes or adata.obs columns)
    use_raw : bool, optional
        Use raw counts (default: False)
    layer : str, optional
        Layer to use for gene expression (default: None)
    output_dir : str or Path, optional
        Output directory for plots (default: ".")
    figsize : tuple, optional
        Figure size (default: auto-calculated)
    ncols : int, optional
        Number of columns in grid (default: 3)
    cmap : str, optional
        Color map (default: 'viridis')

    Returns
    -------
    None
        Saves plots to output_dir
    """
    import scanpy as sc

    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    if 'X_umap' not in adata.obsm:
        print("Warning: UMAP not found. Run run_umap_reduction first.")
        return

    # Filter features that exist
    valid_features = []
    for feature in features:
        if feature in adata.var_names or feature in adata.obs.columns:
            valid_features.append(feature)
        else:
            print(f"Warning: {feature} not found, skipping")

    if len(valid_features) == 0:
        print("No valid features to plot")
        return

    print(f"Plotting UMAP for {len(valid_features)} features...")

    sc.pl.umap(
        adata,
        color=valid_features,
        use_raw=use_raw,
        layer=layer,
        ncols=ncols,
        cmap=cmap,
        frameon=False,
        show=False
    )

    output_file = output_dir / "umap_features"
    fig = plt.gcf()
    _save_plot(fig, output_file, dpi=300)
    plt.close()


def plot_umap_with_plotnine(
    adata: 'AnnData',
    color_by: str,
    output_dir: Union[str, Path] = ".",
    figsize: tuple = (8, 6),
    point_size: float = 0.5
) -> None:
    """
    Create UMAP plot using plotnine with Prism theme.

    Parameters
    ----------
    adata : AnnData
        AnnData object with UMAP
    color_by : str
        Column to color by (from adata.obs)
    output_dir : str or Path, optional
        Output directory for plot (default: ".")
    figsize : tuple, optional
        Figure size in inches (default: (8, 6))
    point_size : float, optional
        Point size (default: 0.5)

    Returns
    -------
    None
        Saves plot to output_dir
    """
    try:
        from plotnine import ggplot, aes, geom_point, labs, theme
        from plotnine_prism import theme_prism
    except ImportError:
        print("Warning: plotnine or plotnine_prism not installed. Using scanpy plotting instead.")
        plot_umap_clusters(adata, cluster_key=color_by, output_dir=output_dir)
        return

    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    if 'X_umap' not in adata.obsm:
        print("Warning: UMAP not found. Run run_umap_reduction first.")
        return

    if color_by not in adata.obs.columns:
        print(f"Warning: {color_by} not found in adata.obs")
        return

    print(f"Plotting UMAP with plotnine (colored by {color_by})...")

    # Prepare data
    umap_df = pd.DataFrame(
        adata.obsm['X_umap'],
        columns=['UMAP1', 'UMAP2']
    )
    umap_df[color_by] = adata.obs[color_by].values

    # Create plot
    plot = (
        ggplot(umap_df, aes(x='UMAP1', y='UMAP2', color=color_by))
        + geom_point(size=point_size, alpha=0.7)
        + labs(title=f'UMAP colored by {color_by}')
        + theme_prism()
    )

    # Save plot
    output_file = output_dir / f"umap_{color_by}_plotnine.svg"
    plot.save(output_file, dpi=300, width=figsize[0], height=figsize[1])

    print(f"  Saved: {output_file}")


def plot_embedding_density(
    adata: 'AnnData',
    groupby: str,
    embedding: str = 'X_umap',
    output_dir: Union[str, Path] = ".",
    figsize: tuple = (12, 4)
) -> None:
    """
    Plot density of cells in embedding space by group.

    Parameters
    ----------
    adata : AnnData
        AnnData object with embedding
    groupby : str
        Column to group by (from adata.obs)
    embedding : str, optional
        Embedding to use (default: 'X_umap')
    output_dir : str or Path, optional
        Output directory for plot (default: ".")
    figsize : tuple, optional
        Figure size (default: (12, 4))

    Returns
    -------
    None
        Saves plot to output_dir
    """
    import scanpy as sc

    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    if embedding not in adata.obsm:
        print(f"Warning: {embedding} not found in adata.obsm")
        return

    if groupby not in adata.obs.columns:
        print(f"Warning: {groupby} not found in adata.obs")
        return

    print(f"Plotting embedding density for {groupby}...")

    sc.pl.embedding_density(
        adata,
        basis=embedding.replace('X_', ''),
        key=groupby,
        show=False
    )

    output_file = output_dir / f"{embedding.replace('X_', '')}_density_{groupby}"
    fig = plt.gcf()
    _save_plot(fig, output_file, dpi=300)
    plt.close()

File: data/skills/scrnaseq-seurat-core-analysis/scripts/plot_dimreduction.R
# ============================================================================
# DIMENSIONALITY REDUCTION PLOTTING
# ============================================================================
#
# Visualization of PCA, UMAP, and clustering results.
#
# Functions:
#   - plot_umap_clusters(): UMAP colored by clusters
#   - plot_clustering_comparison(): Compare multiple resolutions
#   - plot_feature_umap(): Feature expression on UMAP
#   - plot_pca_clusters(): PCA colored by clusters
#
# Usage:
#   source("scripts/plot_dimreduction.R")
#   plot_umap_clusters(seurat_obj, output_dir = "results/umap")

library(ggplot2)
library(ggprism)
library(patchwork)

# Try to load svglite for high-quality SVG (optional)
.has_svglite <- requireNamespace("svglite", quietly = TRUE)
if (.has_svglite) {
    library(svglite)
}

# Internal helper: Save plot in both PNG and SVG formats with graceful fallback
.save_plot <- function(plot, base_path, width = 8, height = 6, dpi = 300) {
    # Always save PNG
    png_path <- sub("\\.(svg|png)$", ".png", base_path)
    ggsave(png_path, plot = plot, width = width, height = height, dpi = dpi, device = "png")
    cat("   Saved:", png_path, "\n")

    # Always try SVG - try ggsave first, fall back to svg() device
    svg_path <- sub("\\.(svg|png)$", ".svg", base_path)
    tryCatch({
        ggsave(svg_path, plot = plot, width = width, height = height, device = "svg")
        cat("   Saved:", svg_path, "\n")
    }, error = function(e) {
        # If ggsave fails, try base R svg() device directly
        tryCatch({
            svg(svg_path, width = width, height = height)
            print(plot)
            dev.off()
            cat("   Saved:", svg_path, "\n")
        }, error = function(e2) {
            cat("   (SVG export failed)\n")
        })
    })
}

#' Plot UMAP colored by clusters
#'
#' @param seurat_obj Seurat object with UMAP and clusters
#' @param group_by Metadata column to color by (default: active idents)
#' @param label Label clusters on plot (default: TRUE)
#' @param label_size Size of labels (default: 6)
#' @param output_dir Output directory for plots
#' @param width Plot width (default: 8)
#' @param height Plot height (default: 6)
#' @return ggplot object
#' @export
plot_umap_clusters <- function(seurat_obj,
                               group_by = NULL,
                               label = TRUE,
                               label_size = 6,
                               output_dir = NULL,
                               width = 8,
                               height = 6) {

  message("Creating UMAP cluster plot")

  # Create UMAP plot
  p <- DimPlot(
    seurat_obj,
    reduction = "umap",
    group.by = group_by,
    label = label,
    label.size = label_size,
    pt.size = 0.5
  ) +
    theme_prism() +
    theme(legend.position = "right") +
    labs(title = "UMAP of Cell Clusters")

  # Save plot if output directory specified
  if (!is.null(output_dir)) {
    dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
    .save_plot(p, file.path(output_dir, "umap_clusters.png"), width, height, dpi = 300)
  }

  return(p)
}

#' Compare clustering at multiple resolutions
#'
#' @param seurat_obj Seurat object with multiple resolution columns
#' @param resolutions Resolutions to compare
#' @param output_dir Output directory for plots
#' @param width Plot width (default: 16)
#' @param height Plot height (default: 12)
#' @return Combined ggplot object
#' @export
plot_clustering_comparison <- function(seurat_obj,
                                      resolutions = c(0.4, 0.6, 0.8, 1.0),
                                      output_dir = NULL,
                                      width = 16,
                                      height = 12) {

  message("Comparing clustering resolutions")

  plots <- list()

  for (res in resolutions) {
    col_name <- paste0("RNA_snn_res.", res)

    if (!col_name %in% colnames(seurat_obj@meta.data)) {
      warning("Resolution ", res, " not found. Skipping.")
      next
    }

    p <- DimPlot(
      seurat_obj,
      reduction = "umap",
      group.by = col_name,
      label = TRUE,
      pt.size = 0.5
    ) +
      theme_prism() +
      labs(title = paste0("Resolution ", res)) +
      theme(legend.position = "none")

    plots[[as.character(res)]] <- p
  }

  # Combine plots
  combined <- wrap_plots(plots, ncol = 2)

  # Save plot if output directory specified
  if (!is.null(output_dir)) {
    dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
    .save_plot(combined, file.path(output_dir, "clustering_comparison.png"), width, height, dpi = 300)
  }

  return(combined)
}

#' Plot feature expression on UMAP
#'
#' @param seurat_obj Seurat object with UMAP
#' @param features Vector of features to plot (genes or metadata columns)
#' @param ncol Number of columns in plot layout (default: 2)
#' @param output_dir Output directory for plots
#' @param width Plot width (default: 12)
#' @param height Plot height (default: calculated from features)
#' @return ggplot object
#' @export
plot_feature_umap <- function(seurat_obj,
                              features,
                              ncol = 2,
                              output_dir = NULL,
                              width = 12,
                              height = NULL) {

  message("Creating feature plots for: ", paste(features, collapse = ", "))

  # Calculate height if not specified
  if (is.null(height)) {
    nrow <- ceiling(length(features) / ncol)
    height <- nrow * 4
  }

  # Create feature plot
  p <- FeaturePlot(
    seurat_obj,
    features = features,
    ncol = ncol,
    pt.size = 0.5,
    reduction = "umap"
  ) &
    theme_prism() &
    scale_color_viridis_c()

  # Save plot if output directory specified
  if (!is.null(output_dir)) {
    dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

    # Create filename from features
    filename <- paste0("feature_plot_", paste(gsub("[^[:alnum:]]", "", features[1:min(3, length(features))]), collapse = "_"))

    .save_plot(p, file.path(output_dir, paste0(filename, ".png")), width, height, dpi = 300)
  }

  return(p)
}

#' Plot PCA colored by clusters
#'
#' @param seurat_obj Seurat object with PCA and clusters
#' @param dims PCs to plot (default: c(1, 2))
#' @param group_by Metadata column to color by (default: active idents)
#' @param output_dir Output directory for plots
#' @param width Plot width (default: 8)
#' @param height Plot height (default: 6)
#' @return ggplot object
#' @export
plot_pca_clusters <- function(seurat_obj,
                              dims = c(1, 2),
                              group_by = NULL,
                              output_dir = NULL,
                              width = 8,
                              height = 6) {

  message("Creating PCA plot")

  # Create PCA plot
  p <- DimPlot(
    seurat_obj,
    reduction = "pca",
    dims = dims,
    group.by = group_by,
    pt.size = 0.5
  ) +
    theme_prism() +
    theme(legend.position = "right") +
    labs(title = paste0("PCA: PC", dims[1], " vs PC", dims[2]))

  # Save plot if output directory specified
  if (!is.null(output_dir)) {
    dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
    .save_plot(p, file.path(output_dir, paste0("pca_PC", dims[1], "_vs_PC", dims[2], ".png")), width, height, dpi = 300)
  }

  return(p)
}

#' Create split UMAP by condition
#'
#' @param seurat_obj Seurat object with UMAP
#' @param split_by Metadata column to split by
#' @param ncol Number of columns (default: auto)
#' @param output_dir Output directory for plots
#' @param width Plot width (default: 12)
#' @param height Plot height (default: 6)
#' @return ggplot object
#' @export
plot_split_umap <- function(seurat_obj,
                            split_by,
                            ncol = NULL,
                            output_dir = NULL,
                            width = 12,
                            height = 6) {

  message("Creating split UMAP by: ", split_by)

  # Create split UMAP
  p <- DimPlot(
    seurat_obj,
    reduction = "umap",
    split.by = split_by,
    ncol = ncol,
    pt.size = 0.5
  ) +
    theme_prism()

  # Save plot if output directory specified
  if (!is.null(output_dir)) {
    dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
    .save_plot(p, file.path(output_dir, paste0("umap_split_by_", split_by, ".png")), width, height, dpi = 300)
  }

  return(p)
}
